index.js

```diff
@@ -291,3 +292,48 @@ function getLinks (entry, decoders) {
   // raw! no links!
   return []
 }
+
+/**
+ * Create a depth-first search function.
+ * Call it with the latest links, it returns the link(s) to follow next.
+ * Maintains a queue of links it has seen but not offered up yet.
+ *
+ * In depth first, we have to resolve links one at a time; we have to
+ * find out if there are child links to follow before trying siblings
+ *
+ * e.g.
+ *
+ * o
+ * ├── x
+ * │   ├── x1
+ * │   └── x2
+ * ├── y
+ * └── z
+ *     └── z1
+ *
+ * [x, y, z] => [x]   (queue: [y, z])
+ *  [x1, x2] => [x1]  (queue: [x2, y, z])
+ *        [] => [x2]  (queue: [y, z])
+ *        [] => [y]   (queue: [z])
+ *        [] => [z]   (queue: [])
+ *      [z1] => [z1]  (queue: [])
+ */
+export function depthFirst () {
+  /** @type {import('multiformats').UnknownLink[]} */
+  let queue = []
+
+  /** @param {import('multiformats').UnknownLink[]} links */
+  return (links = []) => {
+    queue = links.concat(queue)
```

Comment:
🤷‍♂️ Might be nicer to `shift` and `unshift`.

```suggestion
  const queue = []

  /** @param {import('multiformats').UnknownLink[]} links */
  return (links = []) => {
    queue.unshift(...links)
```

Reply:
i dunno, i prefer let & concat here, than spread & unshift